src/compiler/checker.ts

```diff
@@ -23756,6 +23756,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return resolved.callSignatures.length === 0 && resolved.constructSignatures.length === 0 && resolved.indexInfos.length === 0 &&
                 resolved.properties.length > 0 && every(resolved.properties, p => !!(p.flags & SymbolFlags.Optional));
         }
+        if (isNoInferType(type)) {

```

Comment:
I think it would be better to erase all substitution types (to their base types) instead of just `NoInfer` types.

Reply:
I pushed out the requested change but without any new tests.

I've tried a bunch of things based on my intuition in an attempt to write some tests for this change and nothing - from my attempts - managed to hit those lines with a non-NoInfer substitution type. So I went to recheck if I could observe some cases like this using the existing test suite:
- `isExcessPropertyCheckTarget`/`isKnownProperty` didn't observe such substitution types at all
- `isWeakType` has observed it in [`deeplyNestedMappedTypes`](https://github.com/microsoft/TypeScript/blob/ede8ad84b1b2708a0eea234babd5bbf89fda57b8/tests/cases/compiler/deeplyNestedMappedTypes.ts) and [`conditionalTypeDoesntSpinForever`](https://github.com/microsoft/TypeScript/blob/ede8ad84b1b2708a0eea234babd5bbf89fda57b8/tests/cases/compiler/conditionalTypeDoesntSpinForever.ts). Those are complex enough that figuring out a test case based solely on them might take me a longer moment.

Reply:
Yeah, it's not going to be easy to come up with a case where it matters and I'm fine going without.